python/graphstorm/gconstruct/transform.py

```diff
@@ -935,6 +910,185 @@ def call(self, feats):
+
+    def pre_process(self, feats):
+        """ Pre-process input feats
+
+            Not all the edges have the same number of hard negatives.
+            Thus we need to know the maxmun number of hard negatives first.
+
+        Parameters
+        ----------
+        feats:
+            feats to be processed
+
+        Return:
+            dict: {feature_name: feats_statistics}
+        """
+        assert isinstance(feats, (np.ndarray, ExtMemArrayWrapper)), \
+            "Feature of HardEdgeNegativeTransform must be numpy array or ExtMemArray"
+
+        if self._separator is None:
+            max_dim = feats.shape[1]
+        else:
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
+            feats = feats.astype(str)
+            max_dim = 0
+            for feat in feats:
+                dim_size = len(feat.split(self._separator))
+                max_dim = dim_size if dim_size > max_dim else max_dim
+        return {self.feat_name: max_dim}
+
+    def update_info(self, info):
+        max_dim = max(info)
+        self._max_dim = max_dim
+
+    def call(self, feats):
+        """ Parse hard negatives as features
+
+        Hard negatives can be stored as string arrays where
+        each string is a node id. For example:
+
+        .. code::
+
+            src | dst | hard_negs
+            s_0 | d_0 | ["h_0", "h_1"]
+            s_1 | d_1 | ["h_2", "h_3"]
+            s_2 | d_2 | ["h_4", ""]
+            s_3 | d_3 | ["h_5", "h_3"]
+            ...
+
+        Or strings with a delimeter to separate node ids.
+        For example:
+
+        .. code::
+
+            src | dst | hard_negs
+            s_0 | d_0 | "h_0;h_1"
+            s_1 | d_1 | "h_2;h_3"
+            s_2 | d_2 | "h_4"
+            s_3 | d_3 | "h_5;h_3"
+            ...
+
+        Parameters
+        ----------
+        feats : np array
+            Data with hard negatives.
+
+        Returns
+        -------
+        np.array
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        nid_map = self._nid_map[self._target_ntype]
+
+        # It is possible that some edges do not
+        # have enough pre-defined hard negatives.
+        # In certain cases, GraphStorm will fill the
+        # un-provided hard negatives with -1s.
+        neg_ids = np.full((len(feats), self._max_dim), -1, dtype=np.int64)
+        if self._separator is None:
```

Comment:
did you test this on HDF5?

Reply:
No. Let me add a test.

Reply:
Encoding string hdf5 is a little bit complex, maybe we can just skip supporting hdf5.